### PR TITLE
Add lisk address validation function - Closes #531

### LINKED ISSUE #531: Add function to check Lisk address
To check address with `L` in the end - throw when it does not exist, check the amount of characters.

### PR DESCRIPTION
**Pull request by Tosch110**
Closes #531

### CODE REVIEW

File: test/transactions/utils/index.js [hunk 1]
```diff
@@ -13,19 +13,35 @@
...
+	prependMinusToPublicKeys,
+	prependPlusToPublicKeys,
 	signTransaction,
 	multiSignTransaction,
 	verifyTransaction,
+	validateAddress,
+	validateKeysgroup,
+	validatePublicKey,
+	validatePublicKeys,
 } from '../../../src/transactions/utils';
```
Comment (willclarktech) [inline test/transactions/utils/index.js]: Nice, but still missing `signRawTransaction` and `wrapTransactionCreator`.

File: src/transactions/utils/validation.js [hunk 2]
```diff
@@ -47,3 +47,19 @@ export const validateKeysgroup = keysgroup => {
 	}
 	return validatePublicKeys(keysgroup);
 };
+
+// Specification: https://github.com/LiskHQ/lisk/blob/development/logic/account.js#L373
+export const validateAddress = address => {
+	if (address.length < 2 || address.length > 22) {
```
Comment (willclarktech) [inline src/transactions/utils/validation.js]: Not every address with length 22 is valid. A thorough check would see if it's in the valid range using bignum.

File: test/transactions/utils/validation.js [hunk 3]
```diff
@@ -186,4 +187,44 @@ describe('public key validation', () => {
...
+	describe('#validateAddress', () => {
+		describe('Given a valid address', () => {
+			const address = '13133549779353512613L';
+
+			it('should return true if valid', () => {
+				return validateAddress(address).should.be.equal(true);
+			});
+		});
+
+		describe('Given a too short address', () => {
```
Comment (willclarktech) [inline test/transactions/utils/validation.js]: `Given an address that is too short`

File: test/transactions/utils/validation.js [hunk 4]
```diff
@@ -186,4 +187,44 @@ describe('public key validation', () => {
...
+			const address = 'L';
+			const error =
+				'Address length does not match requirements. Expected between 2 and 22 characters.';
+
+			it('should throw if invalid', () => {
+				return validateAddress.bind(null, address).should.throw(error);
+			});
+		});
+
+		describe('Given a too long address', () => {
```
Comment (willclarktech) [inline test/transactions/utils/validation.js]: As above

File: test/transactions/utils/validation.js [hunk 5]
```diff
@@ -186,4 +187,44 @@ describe('public key validation', () => {
 			});
 		});
 	});
+
+	describe('#validateAddress', () => {
+		describe('Given a valid address', () => {
+			const address = '13133549779353512613L';
+
+			it('should return true if valid', () => {
```
Comment (willclarktech) [inline test/transactions/utils/validation.js]: `if valid` is covered by the `describe` block so can be removed.

File: test/transactions/utils/validation.js [hunk 6]
```diff
@@ -186,4 +187,44 @@ describe('public key validation', () => {
...
+				return validateAddress(address).should.be.equal(true);
+			});
+		});
+
+		describe('Given a too short address', () => {
+			const address = 'L';
+			const error =
+				'Address length does not match requirements. Expected between 2 and 22 characters.';
+
+			it('should throw if invalid', () => {
```
Comment (willclarktech) [inline test/transactions/utils/validation.js]: As above.

File: src/transactions/utils/validation.js [hunk 7]
```diff
@@ -47,3 +47,19 @@ export const validateKeysgroup = keysgroup => {
 	}
 	return validatePublicKeys(keysgroup);
 };
+
+// Specification: https://github.com/LiskHQ/lisk/blob/development/logic/account.js#L373
```
Comment (willclarktech) [inline src/transactions/utils/validation.js]: `development` is not a stable reference

File: src/transactions/utils/validation.js [hunk 8]
```diff
@@ -47,3 +47,19 @@ export const validateKeysgroup = keysgroup => {
 	}
 	return validatePublicKeys(keysgroup);
 };
+
+// Specification: https://github.com/LiskHQ/lisk/blob/1.0.0/logic/account.js#L373
```
Comment (willclarktech) [inline src/transactions/utils/validation.js]: `1.0.0` is an even less stable reference than `development`. If we're going to refer to a code line it should be via a commit. Preferably we'd have a documentation site though.

File: src/transactions/utils/validation.js [hunk 9]
```diff
@@ -47,3 +47,19 @@ export const validateKeysgroup = keysgroup => {
...
+export const validateAddress = address => {
+	if (address.length < 2 || address.length > 22) {
```
Comment (willclarktech) [inline src/transactions/utils/validation.js]: This check still doesn't cover the correct numerical range.

File: test/transactions/utils/validation.js [hunk 10]
```diff
@@ -186,4 +187,44 @@ describe('public key validation', () => {
 			});
 		});
 	});
+
+	describe('#validateAddress', () => {
+		describe('Given a valid address', () => {
+			const address = '13133549779353512613L';
+
+			it('should return true', () => {
+				return validateAddress(address).should.be.equal(true);
```
Comment (willclarktech) [inline test/transactions/utils/validation.js]: Prefer `.should.be.true()`

File: src/transactions/utils/validation.js (hunk 8 above)
Comment (Tosch110) [reply src/transactions/utils/validation.js]: Okay, removing this entirely

File: src/transactions/utils/validation.js [hunk 11]
```diff
@@ -61,5 +61,17 @@ export const validateAddress = address => {
 			'Address format does not match requirements. Expected "L" at the end.',
 		);
 	}
+
+	const addressAsBignum = bignum(address.slice(0, -1));
+	const biggestAddress = bignum
+		.fromBuffer(Buffer.from([255, 255, 255, 255, 255, 255, 255, 255]))
+		.plus(1);
```
Comment (willclarktech) [inline src/transactions/utils/validation.js]: Why `plus(1)`?

File: src/transactions/utils/validation.js [hunk 12]
```diff
@@ -61,5 +61,17 @@ export const validateAddress = address => {
 			'Address format does not match requirements. Expected "L" at the end.',
 		);
 	}
+
+	const addressAsBignum = bignum(address.slice(0, -1));
+	const biggestAddress = bignum
+		.fromBuffer(Buffer.from([255, 255, 255, 255, 255, 255, 255, 255]))
```
Comment (willclarktech) [inline src/transactions/utils/validation.js]: How about `new Array(8).fill(255)` so we don't have to count?

File: src/transactions/utils/validation.js [hunk 13]
```diff
@@ -61,5 +61,17 @@ export const validateAddress = address => {
...
+		.plus(1);
+
+	if (addressAsBignum.gt(biggestAddress)) {
+		throw new Error(
+			'Address length does not match requirements. Address out of maximum range.',
```
Comment (willclarktech) [inline src/transactions/utils/validation.js]: It's not really about length in this case.

File: src/transactions/utils/validation.js (hunk 11 above)
Comment (Tosch110) [reply src/transactions/utils/validation.js]: Shouldnt the biggest address still be a possible address?

File: src/transactions/utils/validation.js [hunk 14]
```diff
@@ -63,13 +63,11 @@ export const validateAddress = address => {
 	}
 
 	const addressAsBignum = bignum(address.slice(0, -1));
-	const biggestAddress = bignum
-		.fromBuffer(Buffer.from([255, 255, 255, 255, 255, 255, 255, 255]))
-		.plus(1);
+	const biggestAddress = bignum.fromBuffer(Buffer.from(new Array(8).fill(255)));
```
Comment (willclarktech) [inline src/transactions/utils/validation.js]: How about defining this in `constants`?

Comment (Tosch110) [reply src/transactions/utils/validation.js]: Yeah, I also had that thought. Lets do that then.

File: src/constants/index.js [hunk 15]
```diff
@@ -34,3 +34,6 @@ export const SSL_PORT = '443';
 
 export const GET = 'GET';
 export const POST = 'POST';
+
+// biggesr address derived from bignum.fromBuffer(Buffer.from(new Array(8).fill(255)))
```
Comment (willclarktech) [inline src/constants/index.js]: "Largest possible address. Derived from `bignum.fromBuffer(Buffer.from(new Array(8).fill(255)))`."

**Pull request merged**